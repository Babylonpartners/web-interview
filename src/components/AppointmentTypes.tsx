import * as React from 'react'
import { activeButtonStyles, headerStyles } from '../styles/styles';
import { Responsive } from 'semantic-ui-react';

interface Props {
    selectedAppointmentType: any
    selectedConsultantType: any
    handleChange: any
    availableSlots: any
}

export class AppointmentTypes extends React.PureComponent<Props> {
    render() {

          // calculate slot appointment types

          let appTypeArray = []
          let slotData = this.props.availableSlots
          let appTypeData = this.props.availableSlots

          for (let i = 0, len=appTypeData.length ; i < len; i++) {
            let appointmentType = slotData[i]['appointmentType' as any]
            appTypeArray.push(appointmentType)
          }
          let appTypeMerged = [].concat.apply([], appTypeArray as any)
          const uniqueAppointmentTypes = Array.from(new Set(appTypeMerged))

          if(uniqueAppointmentTypes.length < 1) return <p data-testid="appointmentTypesLoading">Loading...</p>
          return (

          <div data-testid="appointmentTypesComponent">
            
            
            <Responsive maxWidth={640}>
            <strong
              style={headerStyles}
            >
                Appointment Type
            </strong>
            <br />
            <br />
            <>
            {this.props.selectedConsultantType === ""
            ?
            null
            :
            <>
            {uniqueAppointmentTypes.map((appointmentType: any, i: any) => (
              <React.Fragment
                key={i}

              >
              {this.props.selectedAppointmentType === appointmentType
              ?
              <button
                data-testid="selectedAppTypeButton"
                style={activeButtonStyles}
                className="mobile-button"
                value={appointmentType}
                onClick={this.props.handleChange}
              >
               {appointmentType.charAt(0).toUpperCase() + appointmentType.slice(1)}  
              </button>
              :
              <button
                data-testid="unselectedAppTypeButton"
                style={activeButtonStyles}
                value={appointmentType}
                className="mobile-white-button"
                onClick={this.props.handleChange}
              >
                {appointmentType.charAt(0).toUpperCase() + appointmentType.slice(1)}  
              </button>
              }
              </React.Fragment>
            ))}
            </>
            }
            </>
            </Responsive>



            <Responsive minWidth={641}>
            <strong
              style={headerStyles}
            >
                Appointment Type
            </strong>
            <br />
            <br />
            <>
            {this.props.selectedConsultantType === ""
            ?
            null
            :
            <>
            {uniqueAppointmentTypes.map((appointmentType: any, i: any) => (
              <React.Fragment
                key={i}
              >
              {this.props.selectedAppointmentType === appointmentType
              ?
              <button
                data-testid="selectedAppTypeButton"
                style={activeButtonStyles}
                className="button"
                value={appointmentType}
                onClick={this.props.handleChange}
              >
               {appointmentType.charAt(0).toUpperCase() + appointmentType.slice(1)}  
              </button>
              :
              <button
                data-testid="unselectedAppTypeButton"
                style={activeButtonStyles}
                value={appointmentType}
                className="white-button"
                onClick={this.props.handleChange}
              >
                {appointmentType.charAt(0).toUpperCase() + appointmentType.slice(1)}  
              </button>
              }
              </React.Fragment>
            ))}
            </>
            }
            </>
            </Responsive>


          </div>

        );
}}