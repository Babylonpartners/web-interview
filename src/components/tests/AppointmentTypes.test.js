import React from 'react'
import { render, cleanup } from '@testing-library/react'
import { AppointmentTypes } from '../AppointmentTypes'

afterEach(() => {
  cleanup()
  console.error.mockClear()
})

console.error = jest.fn()

const slots = [
  {
    id: 1,
    consultantType: ['gp'],
    appointmentType: ['audio', 'video'],
    time: '2019-11-27T10:11:00.000Z',
  },
  {
    id: 2,
    consultantType: ['specialist', 'gp'],
    appointmentType: ['audio', 'video'],
    time: '2019-12-01T14:16:30.000Z',
  },
]

test('<AppointmentTypes /> slot data NOT recieved', () => {
  const { getByTestId, queryByTestId, container } = render(
    <AppointmentTypes
      selectedAppointmentType
      selectedConsultantType
      handleChange
      availableSlots
    />
  )
  expect(container.firstChild).toMatchSnapshot()
  expect(getByTestId('appointmentTypesLoading')).toBeTruthy()
  expect(queryByTestId('appointmentTypesComponent')).toBeFalsy()
})

test('<AppointmentTypes /> slot data NOT recieved', () => {
  const { getByTestId, queryByTestId } = render(
    <AppointmentTypes
      selectedAppointmentType
      selectedConsultantType
      handleChange
      availableSlots={slots}
    />
  )

  expect(queryByTestId('appointmentTypesLoading')).toBeFalsy()
  expect(getByTestId('appointmentTypesComponent')).toBeTruthy()
})

test('<AppointmentTypes /> number items unselected', () => {
  const { getAllByTestId } = render(
    <AppointmentTypes
      handleChange
      selectedConsultantType={'gp'}
      selectedAppointmentType
      availableSlots={slots}
    />
  )
  expect(getAllByTestId('unselectedAppTypeButton').length).toBe(2)
})

test('<AppointmentTimes /> 1 button selected', () => {
  const { getAllByTestId } = render(
    <AppointmentTypes
      handleChange
      selectedConsultantType={'gp'}
      selectedAppointmentType={'audio'}
      availableSlots={slots}
    />
  )
  expect(getAllByTestId('selectedAppTypeButton').length).toBe(1)
})
