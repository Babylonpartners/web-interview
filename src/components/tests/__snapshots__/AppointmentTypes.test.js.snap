// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppointmentTypes /> slot data NOT recieved 1`] = `
<p
  data-testid="appointmentTypesLoading"
>
  Loading...
</p>
`;
